examples/pointer.c

```diff
@@ -42,112 +54,156 @@ static void handle_output_frame(struct output_state *output, struct timespec *ts
-		sample->default_color[i] += delta > 0 ? -0.05f : 0.05f;
+		sample->default_color[i] += event->delta > 0 ? -0.05f : 0.05f;
 		if (sample->default_color[i] > 1.0f) {
 			sample->default_color[i] = 1.0f;
 		}
 		if (sample->default_color[i] < 0.0f) {
 			sample->default_color[i] = 0.0f;
 		}
 	}
+
 	memcpy(&sample->clear_color, &sample->default_color,
 			sizeof(sample->clear_color));
 }
 
-static void handle_output_add(struct output_state *ostate) {
-	struct sample_state *state = ostate->compositor->data;
-	struct wlr_output *wlr_output = ostate->output;
-	struct wlr_xcursor_image *image = state->cursor->images[0];
-	if (!wlr_output_set_cursor(wlr_output, image->buffer,
-			image->width, image->width, image->height)) {
-		wlr_log(L_DEBUG, "Failed to set hardware cursor");
-		return;
-	}
-	if (!wlr_output_move_cursor(wlr_output, 0, 0)) {
-		wlr_log(L_DEBUG, "Failed to move hardware cursor");
-	}
-}
-
 int main(int argc, char *argv[]) {
 	struct sample_state state = {
 		.default_color = { 0.25f, 0.25f, 0.25f, 1 },
 		.clear_color = { 0.25f, 0.25f, 0.25f, 1 }
 	};
+
+	state.config = parse_args(argc, argv);
+	state.cursor = wlr_cursor_init();
+
+	wl_signal_add(&state.cursor->events.motion, &state.cursor_motion);
+	state.cursor_motion.notify = handle_cursor_motion;
+
+	wl_signal_add(&state.cursor->events.motion_absolute, &state.cursor_motion_absolute);
+	state.cursor_motion_absolute.notify = handle_cursor_motion_absolute;
+
+	wl_signal_add(&state.cursor->events.button, &state.cursor_button);
+	state.cursor_button.notify = handle_cursor_button;
+
+	wl_signal_add(&state.cursor->events.axis, &state.cursor_axis);
+	state.cursor_axis.notify = handle_cursor_axis;
+
 	struct compositor_state compositor = { 0 };
 	compositor.data = &state;
 	compositor.output_add_cb = handle_output_add;
+	compositor.output_remove_cb = handle_output_remove;
+	compositor.output_resolution_cb = handle_output_resolution;
 	compositor.output_frame_cb = handle_output_frame;
-	compositor.pointer_motion_cb = handle_pointer_motion;
-	compositor.pointer_motion_absolute_cb = handle_pointer_motion_absolute;
-	compositor.pointer_button_cb = handle_pointer_button;
-	compositor.pointer_axis_cb = handle_pointer_axis;
+	compositor.input_add_cb = handle_input_add;
+
+	state.compositor = &compositor;
 
 	struct wlr_xcursor_theme *theme = wlr_xcursor_theme_load("default", 16);
 	if (!theme) {
 		wlr_log(L_ERROR, "Failed to load cursor theme");
 		return 1;
 	}
-	state.cursor = wlr_xcursor_theme_get_cursor(theme, "left_ptr");
-	if (!state.cursor) {
+	state.xcursor = wlr_xcursor_theme_get_cursor(theme, "left_ptr");
+	if (!state.xcursor) {
 		wlr_log(L_ERROR, "Failed to load left_ptr cursor");
 		return 1;
 	}
 
+	wlr_cursor_set_xcursor(state.cursor, state.xcursor);
+
 	compositor_init(&compositor);
 	wl_display_run(compositor.display);
 	compositor_fini(&compositor);
 
 	wlr_xcursor_theme_destroy(theme);
+	example_config_destroy(state.config);
+	wlr_cursor_destroy(state.cursor);
```

Comment:
It needed `wl_list_for_each_safe`. You have to use that for all your cleanup loops I guess.